packages/ceramic-core/src/store/level-state-store.ts

```diff
@@ -0,0 +1,217 @@
+import CID from "cids"
+import Level from 'level-ts';
+
+import Ipfs from 'ipfs'
+import Document, { AnchorStatus, DocState } from "../document"
+import StateStore from "./state-store"
+import Dispatcher from "../dispatcher"
+
+/**
+ * Level backed State Store
+ */
+export default class LevelStateStore implements StateStore {
+    private store: any
+
+    constructor(private ipfs: Ipfs.Ipfs, private dispatcher: Dispatcher, private storePath: string) {
+    }
+
+    /**
+     * Open pinning service
+     */
+    async open(): Promise<void> {
+        this.store = new Level(this.storePath);
+    }
+
+    /**
+     * Pin document
+     * @param document - Document instance
+     * @param pinOnIpfs - Pin logs on IPFS
+     */
+    async pin(document: Document, pinOnIpfs = true): Promise<void> {
+        const { state } = document;
+        await this.store.put(document.id, this._serializeState(state))
+
+        if (pinOnIpfs) {
+            const { log } = state;
+            const pinPromises = log.map(async (cid) => {
+                await this.ipfs.pin.add(cid.toString(), {
+                    recursive: false,
+                })
+
+                const record = await this.dispatcher.retrieveRecord(cid)
+                if (record.proof) {
+                    // fetch and pin
+                    await this.dispatcher.retrieveRecord(record.proof) // fetch proof object
+                    await this.ipfs.pin.add(record.proof.toString(), {
+                        recursive: false,
+                    })
+
+                    const path = "root/" + record.path
+                    const subPaths = path.split('/')
+
+                    let currentPath = ""
+                    for (const subPath of subPaths) {
+                        if (subPath.length === 0) {
+                            continue
+                        }
+                        currentPath += "/" + subPath
+
+                        // fetch and pin
+                        await this.ipfs.dag.get(record.proof, currentPath)
```

Comment:
Do we really need to call `dag.get` here? Doesn't `pin.add` do this automatically?

Reply:
I assumed that we need to fetch it. If `pin.add` does that automatically, even better :)